src/Database/LSMTree/Internal/IndexOrdinaryAcc.hs

```diff
@@ -0,0 +1,100 @@
+{- HLINT ignore "Avoid restricted alias" -}
+
+{-|
+    Incremental construction functionality for the general-purpose fence pointer
+    index.
+-}
+module Database.LSMTree.Internal.IndexOrdinaryAcc
+(
+    IndexOrdinaryAcc,
+    new,
+    append,
+    end
+)
+where
+
+import           Control.Monad.ST.Strict (ST)
+import           Data.List (genericReplicate)
+import           Data.Primitive.ByteArray (byteArrayFromListN)
+import           Data.STRef.Strict (STRef, modifySTRef, newSTRef, readSTRef)
+import           Data.Vector (fromList)
+import qualified Data.Vector.Primitive as Primitive (Vector, length)
+import           Data.Word (Word16, Word8)
+import           Database.LSMTree.Internal.Chunk (Baler, Chunk, createBaler,
+                     feedBaler, readBalerRemnant)
+import           Database.LSMTree.Internal.IndexCompactAcc
+                     (Append (AppendMultiPage, AppendSinglePage))
+import           Database.LSMTree.Internal.IndexOrdinary
+                     (IndexOrdinary (IndexOrdinary))
+import           Database.LSMTree.Internal.Serialise
+                     (SerialisedKey (SerialisedKey'))
+import           Database.LSMTree.Internal.Vector (mkPrimVector)
+
+{-|
+    A general-purpose fence pointer index under incremental construction.
+
+    A value @IndexOrdinaryAcc lastKeysRevRef baler@ denotes a partially
+    constructed index with the following properties:
+
+      * The keys pointed to by @lastKeysRevRef@ are the keys that the index
+        assigns to pages, but in reverse order.
+      * The @baler@ object is used by the index for incremental output of the
+        serialised key list.
+-}
+data IndexOrdinaryAcc s = IndexOrdinaryAcc !(STRef s [SerialisedKey]) !(Baler s)
+
+-- | Creates a new, initially empty, index.
+new :: Int                       -- ^ Maximum chunk size in bytes
+    -> ST s (IndexOrdinaryAcc s) -- ^ Construction of the index
+new maxChunkSize = IndexOrdinaryAcc <$> newSTRef [] <*> createBaler maxChunkSize
+
+{-
+    Appends a single key to the key list of an index and outputs newly available
+    chunks of the serialised key list.
+-}
+appendKey :: SerialisedKey -> IndexOrdinaryAcc s -> ST s [Chunk]
+appendKey lastKey@(SerialisedKey' lastKeyBytes)
+          (IndexOrdinaryAcc lastKeysRevRef baler)
+    = do
+          modifySTRef lastKeysRevRef (lastKey :)
+          chunksIntoLastKeySize <- feedBaler lastKeySizeBytes baler
+          chunksIntoLastKey <- feedBaler lastKeyBytes baler
+          return (chunksIntoLastKeySize ++ chunksIntoLastKey)
+    where
+
+    lastKeySize :: Int
+    lastKeySize = Primitive.length lastKeyBytes
+
+    lastKeySizeAsWord16 :: Word16
+    lastKeySizeAsWord16
+        | lastKeySize <= fromIntegral (maxBound :: Word16)
+            = fromIntegral lastKeySize
+        | otherwise
+            = error "Serialised key too large"
```

Comment:
The index does not have to check this. It should be checked by something like the `PageAcc`. https://github.com/IntersectMBO/lsm-tree/issues/304

Reply:
I think that this module should still provide a somewhat safe interface, despite being internal. The `SerialisedKey` type allows serialized keys to have any size that can be represented by an `Int`, so that the interface of this module doesn’t prevent keys from being larger than what can be represented by a `Word16`. What makes matters worse is that the `fromIntegral` implementation for `Word16` computes remainders modulo 2¹⁶, so that, without a check, submitting too large keys will silently corrupt the index, in a somewhat subtle way.

I think that there should at least be an assertion in this place. The API documentation should then clearly state the size constraint for keys, as assertion checking may be switched off.

Reply:
I made a corresponding change in f1e84da3bad7307b0468eee806552dd8cdddb7c6. Please tell me whether you’re happy with it or not. 🙂